pkg/action/hooks.go

```diff
@@ -97,9 +98,11 @@ func (cfg *Configuration) execHook(rl *release.Release, hook release.HookEvent,
 		h.LastRun.Phase = release.HookPhaseSucceeded
 	}
 
+	hooksToDelete := releaseutil.SortHooksToDelete(executingHooks)
```

Comment:
why are we "re-sorting" the hooks here? If the intention is to delete in reverse order, why not just iterate `executingHooks` in reverse order?

Reply:
they should be sorted by kind and by weight, here it sorts only by weight

Reply:
Sorting by type happens when resources are rendered. For example, when they are rendered for install like here... https://github.com/helm/helm/blob/a2cffa6e3f9f921d8083d187d55471615e5b7515/pkg/action/action.go#L168

Those sorted resources are passed into this function (`execHook`). Then, they are sorted again by weight and name. See the implementation at https://github.com/helm/helm/blob/a2cffa6e3f9f921d8083d187d55471615e5b7515/pkg/action/hooks.go#L112-L121.

Any sorting in here again started with resources sorted by type. The difference between sorting the original list again and using the existing list is the part by name.

It should be possible to save on work by iterating the existing sort in reverse order. It would be reverse weight / alpha. It would also save on exposing more of a public API.

Reply:
Maybe I'm missing something, I don't understand why @mattfarina and @gjenkins8 wrote about the initial sorting by type if we rewrite the inital sorting by `hookByWeight`, it doesn't keep the original sorting, it rewrites the sorting order:  

- before `hookByWeight` - the items sorted by type(btw, they will be sorted by install order while we should sort by uninstall order)
- after `hookByWeight` - the items sorted by weight and name

Reply:
I was showing how current sorting works so that people could follow along. Especially some newer maintainers who are on this PR.

What's wrong with doing something like...

```go
for i, := len(executingHooks)-1; i >= 0; i-- {
	if err := cfg.deleteHookByPolicy(executingHooks[i], release.HookSucceeded, timeout); err != nil {
		return err
	}
}
```

This takes the existing `executingHooks` and deals with them in reverse order. Note, I did not test this.

This avoids resorting and just iterates the existing list backwards.

Reply:
@mattfarina we can't just iterate in reverse order, because the hooks are **sorted by weight and name** only, while the hooks should be sorted **by weight and type**

Reply:
@mattfarina hi! check please the comment above

Reply:
@mattfarina because the `executingHooks` are sorted **only**(because it overwrites the sort by kind) by `weight` and if the `weight` is zero(or just the same), then you can't guarantee what order the resources will have.

Reply:
@mattfarina I apologize. I just tried to reproduce the issue that led me to add my custom sorting instead of reversing the order for deletion, but I couldn’t. Either I did something wrong back then, or I don’t remember how to reproduce it. I’ve pushed changes where I simply reverse the order for deletion.